Source/DafnyCore/DooFile.cs

```diff
@@ -0,0 +1,238 @@
+using System;
+using System.Collections.Generic;
+using System.CommandLine;
+using System.Configuration;
+using System.IO;
+using System.IO.Compression;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using Microsoft.Dafny;
+using Microsoft.Dafny.Auditor;
+using Tomlyn;
+using Tomlyn.Model;
+
+namespace DafnyCore; 
+
+// Model class for the .doo file format for Dafny libraries.
+// Contains the validation logic for safely consuming libraries as well.
+public class DooFile {
+
+  private const string ProgramFileEntry = "program.dfy";
+  
+  private const string ManifestFileEntry = "manifest.toml";
+
+  public class ManifestData {
+    public const string CurrentDooFileVersion = "1.0";
+    public string DooFileVersion { get; set; }
+  
+    public string DafnyVersion { get; set; }
+
+    public string SolverIdentifier { get; set; }
+    public string SolverVersion { get; set; }
+    
+    public Dictionary<string, object> Options { get; set; }
+
+    public ManifestData() {
+      // Only for TOML deserialization!
+    }
+    
+    public ManifestData(DafnyOptions options) {
+      DooFileVersion = CurrentDooFileVersion;
+      DafnyVersion = options.VersionNumber;
+
+      SolverIdentifier = options.SolverIdentifier;
+      SolverVersion = options.SolverVersion.ToString();
+      
+      Options = new Dictionary<string, object>();
+      foreach (var (option, check) in OptionChecks) {
+        options.Options.OptionArguments.TryGetValue(option, out var optionValue);
+        Options.Add(option.Name, optionValue);
+      }
+    }
+
+    public static ManifestData Read(TextReader reader) {
+      return Toml.ToModel<ManifestData>(reader.ReadToEnd(), null, new TomlModelOptions());
```

Comment:
Does this check that the text being read only contains entries that fit in `ManifestData`?

Reply:
Yup! Adding a quick unit test for this though. :)